pkg/controller/pod_volume_restore_controller.go

```diff
@@ -261,38 +249,46 @@ func (c *PodVolumeRestoreReconciler) processRestore(ctx context.Context, req *ve
-		if err != nil {
-			log.WithError(err).Error("Error creating temp cacert file")
+	// name of ResticRepository is generated with prefix volumeNamespace-backupLocation- and end with random characters
+	// it could not retrieve the ResticRepository CR with namespace + name. so first list all CRs with in the volumeNamespace
+	// then filtering the matched CR with prefix volumeNamespace-backupLocation-
+	backupRepos := &velerov1api.BackupRepositoryList{}
+	var backupRepo velerov1api.BackupRepository
+	isFoundRepo := false
+	if c.List(ctx, backupRepos, &client.ListOptions{
+		Namespace: req.Namespace,
+	}); err != nil {
+		return errors.Wrap(err, "error getting backup repository")
+	} else if len(backupRepos.Items) == 0 {
+		return errors.Errorf("find empty BackupRepository found for workload namespace %s, backup storage location %s", req.Namespace, req.Spec.BackupStorageLocation)
+	} else {
+		for _, repo := range backupRepos.Items {
+			if strings.HasPrefix(repo.Name, fmt.Sprintf("%s-%s-", req.Spec.Pod.Namespace, req.Spec.BackupStorageLocation)) {
+				backupRepo = repo
+				isFoundRepo = true
+				break
+			}
+		}
+		if !isFoundRepo {
+			return errors.Errorf("could not found match BackupRepository for workload namespace %s, backup storage location %s", req.Namespace, req.Spec.BackupStorageLocation)
 		}
-		// ignore error since there's nothing we can do and it's a temp file.
-		defer os.Remove(caCertFile)
 	}
-	resticCmd.CACertFile = caCertFile
 
-	env, err := restic.CmdEnv(backupLocation, c.credentialsFileStore)
+	uploaderProv, err := provider.NewUploaderProvider(ctx, c.Client, req.Spec.UploaderType,
+		req.Spec.RepoIdentifier, backupLocation, &backupRepo, c.credentialGetter, repokey.RepoKeySelector(), log)
 	if err != nil {
-		return errors.Wrap(err, "error setting restic cmd env")
-	}
-	resticCmd.Env = env
-
-	// #4820: restrieve insecureSkipTLSVerify from BSL configuration for
-	// AWS plugin. If nothing is return, that means insecureSkipTLSVerify
-	// is not enable for Restic command.
-	skipTLSRet := restic.GetInsecureSkipTLSVerifyFromBSL(backupLocation, log)
-	if len(skipTLSRet) > 0 {
-		resticCmd.ExtraFlags = append(resticCmd.ExtraFlags, skipTLSRet)
+		return errors.Wrap(err, "error creating uploader")
 	}
 
-	var stdout, stderr string
+	defer func() {
+		if err := uploaderProv.Close(ctx); err != nil {
+			log.Errorf("failed to close uploader provider with error %v", err)
+		}
+	}()
 
-	if stdout, stderr, err = restic.RunRestore(resticCmd, log, c.updateRestoreProgressFunc(req, log)); err != nil {
-		return errors.Wrapf(err, "error running restic restore, cmd=%s, stdout=%s, stderr=%s", resticCmd.String(), stdout, stderr)
+	if err = uploaderProv.RunRestore(ctx, req.Spec.SnapshotID, volumePath, c.NewRestoreProgressUpdater(req, log, ctx)); err != nil {
+		return errors.Wrapf(err, "error running restic restore err=%v", err)
```

Comment:
"error running restic restore err=%v" to "error running restore err=%v"

Reply:
yeah, I've changed